li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/largemessage/MessageSplitterImpl.java

```diff
@@ -96,8 +108,26 @@ public MessageSplitterImpl(int maxSegmentSize,
 
       // NOTE: Even though we are passing topic here to serialize, the segment itself should be topic independent.
       byte[] segmentValue = _segmentSerializer.serialize(topic, segment);
+
+      //  Make a temporary copy of headers because we'd be overwriting {@link Constants.LARGE_MESSAGE_HEADER}
+      Headers temporaryHeaders = new RecordHeaders(headers);
+      temporaryHeaders.remove(Constants.LARGE_MESSAGE_HEADER);
+      LargeMessageHeaderValue largeMessageHeaderValue = new LargeMessageHeaderValue(
+          LargeMessageHeaderValue.LEGACY,
+          messageId,
+          seq,
+          numberOfSegments
+      );
+
       ProducerRecord<byte[], byte[]> segmentProducerRecord =
-          new ProducerRecord<>(topic, partition, timestamp, segmentKey, segmentValue);
+          new ProducerRecord<>(
+              topic,
+              partition,
+              timestamp,
+              segmentKey,
+              segmentValue,
+              temporaryHeaders.add(Constants.LARGE_MESSAGE_HEADER, LargeMessageHeaderValue.toBytes(largeMessageHeaderValue))
```

Comment:
nit: I suggest calling `add()` followed by `remove()` making it easier to understand that that the large message header is being overwritten.